nibabel/streamlines/compact_list.py

```diff
@@ -0,0 +1,213 @@
+            self._data = np.zeros((0, elem.shape[1]), dtype=elem.dtype)
+
+        if isinstance(elements, CompactList):
+            self._data = np.concatenate([self._data, elements._data], axis=0)
+            lengths = elements._lengths
+        else:
+            self._data = np.concatenate([self._data] + list(elements), axis=0)
+            lengths = list(map(len, elements))
+
+        idx = self._offsets[-1] + self._lengths[-1] if len(self) > 0 else 0
+        self._lengths.extend(lengths)
+        self._offsets.extend(np.cumsum([idx] + lengths).tolist()[:-1])
+
+    def copy(self):
+        """ Creates a copy of this ``CompactList`` object. """
+        # We do not simply deepcopy this object since we might have a chance
+        # to use less memory. For example, if the compact list being copied
+        # is the result of a slicing operation on a compact list.
+        clist = CompactList()
+        total_lengths = np.sum(self._lengths)
+        clist._data = np.empty((total_lengths,) + self._data.shape[1:],
+                               dtype=self._data.dtype)
+
+        idx = 0
+        for offset, length in zip(self._offsets, self._lengths):
+            clist._offsets.append(idx)
+            clist._lengths.append(length)
+            clist._data[idx:idx+length] = self._data[offset:offset+length]
+            idx += length
+
+        return clist
+
+    def __getitem__(self, idx):
+        """ Gets element(s) through indexing.
+
+        Parameters
+        ----------
+        idx : int, slice or list
+            Index of the element(s) to get.
+
+        Returns
+        -------
+        ndarray object(s)
+            When `idx` is a int, returns a single ndarray.
+            When `idx` is either a slice or a list, returns a list of ndarrays.
+        """
+        if isinstance(idx, int) or isinstance(idx, np.integer):
+            start = self._offsets[idx]
+            return self._data[start:start+self._lengths[idx]]
+
+        elif type(idx) is slice:
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = self._offsets[idx]
+            clist._lengths = self._lengths[idx]
+            return clist
+
+        elif type(idx) is list:
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = [self._offsets[i] for i in idx]
+            clist._lengths = [self._lengths[i] for i in idx]
+            return clist
+
+        elif isinstance(idx, np.ndarray) and idx.dtype == np.bool:
+            clist = CompactList()
+            clist._data = self._data
+            clist._offsets = [self._offsets[i]
+                              for i, take_it in enumerate(idx) if take_it]
+            clist._lengths = [self._lengths[i]
+                              for i, take_it in enumerate(idx) if take_it]
+            return clist
+
+        raise TypeError("Index must be a int or a slice! Not " + str(type(idx)))
```

Comment:
But you are right for the mix of `isinstance` and `type`. I'll change the code to use `isinstance` where I can since it is more flexible (supports subclasses too).